// @fluentui-react-native/button/lib-commonjs/StealthButton/StealthButton.settings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.settings = [
    {
        tokens: {
            backgroundColor: 'menuBackground',
            color: 'menuItemText',
            borderColor: 'menuBackground',
        },
        _overrides: {
            disabled: {
                tokens: {
                    color: 'disabledBodyText',
                    borderColor: 'menuBackground',
                    backgroundColor: 'background',
                },
            },
            hovered: {
                tokens: {
                    backgroundColor: 'menuItemBackgroundHovered',
                    color: 'menuItemTextHovered',
                    borderColor: 'menuItemBackgroundHovered',
                },
            },
            pressed: {
                tokens: {
                    backgroundColor: 'menuItemBackgroundPressed',
                    borderColor: 'menuItemBackgroundPressed',
                },
            },
            focused: {
                tokens: {
                    borderColor: 'focusBorder',
                    backgroundColor: 'menuItemBackgroundHovered',
                    color: 'menuItemTextHovered',
                },
            },
        },
    },
    'StealthButton',
];
