// @fluentui-react-native/button/lib-commonjs/Button.settings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Button_types_1 = require("./Button.types");
exports.settings = [
    {
        tokens: {
            backgroundColor: 'buttonBackground',
            color: 'buttonText',
            borderColor: 'buttonBorder',
            borderWidth: 1,
            borderRadius: 4,
        },
        root: {
            accessible: true,
            focusable: true,
            accessibilityRole: 'button',
            style: {
                display: 'flex',
                alignItems: 'flex-start',
                flexDirection: 'row',
                alignSelf: 'flex-start',
            },
        },
        endIcon: {
            style: {
                marginStart: 2,
            },
        },
        startIcon: {
            style: {
                marginEnd: 2,
            },
        },
        content: {
            style: {
                marginStart: 2,
                marginEnd: 2,
            },
        },
        stack: {
            style: {
                display: 'flex',
                paddingStart: 6,
                paddingEnd: 6,
                alignItems: 'center',
                flexDirection: 'row',
                alignSelf: 'flex-start',
                minHeight: 24,
                minWidth: 32,
                justifyContent: 'center',
            },
        },
        _precedence: ['hovered', 'focused', 'pressed', 'disabled'],
        _overrides: {
            disabled: {
                tokens: {
                    backgroundColor: 'buttonBackgroundDisabled',
                    color: 'buttonTextDisabled',
                    borderColor: 'buttonBorderDisabled',
                },
            },
            hovered: {
                tokens: {
                    backgroundColor: 'buttonBackgroundHovered',
                    color: 'buttonTextHovered',
                    borderColor: 'buttonBorderHovered',
                },
            },
            pressed: {
                tokens: {
                    backgroundColor: 'buttonBackgroundPressed',
                    color: 'buttonTextPressed',
                    borderColor: 'buttonPressedBorder',
                },
            },
            focused: {
                tokens: {
                    borderColor: 'buttonBorderFocused',
                    color: 'buttonTextHovered',
                    backgroundColor: 'buttonBackgroundHovered',
                },
            },
        },
    },
    Button_types_1.buttonName,
];
