// @fluentui-react-native/button/lib-commonjs/Button.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
/** @jsx withSlots */
var React = __importStar(require("react"));
var react_native_1 = require("react-native");
var Button_types_1 = require("./Button.types");
var foundation_compose_1 = require("@uifabricshared/foundation-compose");
var foundation_composable_1 = require("@uifabricshared/foundation-composable");
var text_1 = require("@fluentui-react-native/text");
var Button_settings_1 = require("./Button.settings");
var tokens_1 = require("@fluentui-react-native/tokens");
var adapters_1 = require("@fluentui-react-native/adapters");
var foundation_settings_1 = require("@uifabricshared/foundation-settings");
var interactive_hooks_1 = require("@fluentui-react-native/interactive-hooks");
var icon_1 = require("@fluentui-react-native/icon");
exports.Button = foundation_compose_1.compose({
    displayName: Button_types_1.buttonName,
    usePrepareProps: function (userProps, useStyling) {
        var defaultComponentRef = React.useRef(null);
        var icon = userProps.icon, startIcon = userProps.startIcon, endIcon = userProps.endIcon, content = userProps.content, _a = userProps.onAccessibilityTap, onAccessibilityTap = _a === void 0 ? userProps.onClick : _a, _b = userProps.accessibilityLabel, accessibilityLabel = _b === void 0 ? userProps.content : _b, _c = userProps.componentRef, componentRef = _c === void 0 ? defaultComponentRef : _c, testID = userProps.testID, onClick = userProps.onClick, rest = __rest(userProps, ["icon", "startIcon", "endIcon", "content", "onAccessibilityTap", "accessibilityLabel", "componentRef", "testID", "onClick"]);
        // Ensure focus is placed on button after click
        var onPressWithFocus = interactive_hooks_1.useOnPressWithFocus(componentRef, onClick);
        // attach the pressable state handlers
        var pressable = interactive_hooks_1.useAsPressable(__assign(__assign({}, rest), { onPress: onPressWithFocus }));
        var onKeyUp = interactive_hooks_1.useKeyCallback(onClick, ' ', 'Enter');
        // set up state
        var state = {
            info: __assign(__assign({}, pressable.state), { disabled: !!userProps.disabled, content: !!content, startIcon: !!startIcon || !!icon, endIcon: !!endIcon }),
        };
        var buttonRef = interactive_hooks_1.useViewCommandFocus(componentRef);
        // grab the styling information, referencing the state as well as the props
        var styleProps = useStyling(userProps, function (override) { return state.info[override] || userProps[override]; });
        // create the merged slot props
        var slotProps = foundation_settings_1.mergeSettings(styleProps, {
            root: __assign(__assign({}, pressable.props), { ref: buttonRef, onAccessibilityTap: onAccessibilityTap, accessibilityLabel: accessibilityLabel, accessibilityState: { disabled: state.info.disabled }, onKeyUp: onKeyUp }),
            content: { children: content, testID: testID },
            startIcon: interactive_hooks_1.createIconProps(startIcon || icon),
            endIcon: interactive_hooks_1.createIconProps(endIcon),
        });
        return { slotProps: slotProps, state: state };
    },
    settings: Button_settings_1.settings,
    render: function (Slots, renderData) {
        var children = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            children[_i - 2] = arguments[_i];
        }
        var info = renderData.state.info;
        return (foundation_composable_1.withSlots(Slots.root, null,
            foundation_composable_1.withSlots(Slots.borderWrapper, null,
                foundation_composable_1.withSlots(Slots.stack, null,
                    info.startIcon && foundation_composable_1.withSlots(Slots.startIcon, null),
                    info.content && foundation_composable_1.withSlots(Slots.content, null),
                    children,
                    info.endIcon && foundation_composable_1.withSlots(Slots.endIcon, null)))));
    },
    slots: {
        root: react_native_1.View,
        stack: { slotType: react_native_1.View, filter: adapters_1.filterViewProps },
        borderWrapper: { slotType: react_native_1.View, filter: adapters_1.filterViewProps },
        startIcon: { slotType: icon_1.Icon },
        content: text_1.Text,
        endIcon: { slotType: icon_1.Icon },
    },
    styles: {
        root: [tokens_1.backgroundColorTokens, tokens_1.borderTokens],
        stack: [],
        borderWrapper: [{ source: 'wrapperBorderColor', lookup: tokens_1.getPaletteFromTheme, target: 'borderColor' }],
        startIcon: [{ source: 'iconColor', lookup: tokens_1.getPaletteFromTheme, target: 'color' }],
        content: [tokens_1.textTokens, tokens_1.foregroundColorTokens],
        endIcon: [{ source: 'iconColor', lookup: tokens_1.getPaletteFromTheme, target: 'color' }],
    },
});
exports.default = exports.Button;
