// @fluentui-react-native/button/lib-commonjs/Button.settings.android.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Button_types_1 = require("./Button.types");
exports.settings = [
    {
        tokens: {
            backgroundColor: 'buttonBackground',
            color: 'buttonText',
            borderColor: 'buttonBorder',
            borderWidth: 1,
            borderRadius: 12,
        },
        root: {
            accessible: false,
            focusable: false,
            style: {
                display: 'flex',
                alignItems: 'flex-start',
                flexDirection: 'row',
                alignSelf: 'flex-start',
                overflow: 'hidden',
            },
        },
        ripple: {
            accessible: true,
            focusable: true,
            accessibilityRole: 'button',
            android_ripple: {
                color: 'buttonBackgroundPressed',
            },
        },
        content: {
            variant: 'bodySemibold',
        },
        startIcon: {
            style: {
                marginEnd: 10,
            },
        },
        stack: {
            style: {
                display: 'flex',
                paddingVertical: 14,
                paddingHorizontal: 20,
                alignItems: 'center',
                flexDirection: 'row',
                alignSelf: 'flex-start',
                minHeight: 52,
                minWidth: 80,
                justifyContent: 'center',
            },
        },
        _precedence: ['hovered', 'focused', 'pressed', 'disabled'],
        _overrides: {
            disabled: {
                tokens: {
                    backgroundColor: 'buttonBackgroundDisabled',
                    color: 'buttonTextDisabled',
                    borderColor: 'buttonBorderDisabled',
                },
            },
            hovered: {
                tokens: {
                    backgroundColor: 'buttonBackgroundHovered',
                    color: 'buttonTextHovered',
                    borderColor: 'buttonBorderHovered',
                },
            },
            pressed: {
                tokens: {
                    backgroundColor: 'buttonBackgroundPressed',
                    color: 'buttonTextPressed',
                    borderColor: 'buttonPressedBorder',
                },
            },
            focused: {
                tokens: {
                    borderColor: 'buttonBorderFocused',
                    color: 'buttonTextHovered',
                    backgroundColor: 'buttonBackgroundHovered',
                },
            },
        },
    },
    Button_types_1.buttonName,
];
