// @fluentui-react-native/separator/lib-commonjs/Separator.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
/** @jsx withSlots */
var react_native_1 = require("react-native");
var framework_1 = require("@fluentui-react-native/framework");
exports.separatorName = 'Separator';
var propMask = { vertical: undefined };
exports.Separator = framework_1.compose({
    displayName: exports.separatorName,
    tokens: [{ separatorWidth: 1 }, exports.separatorName],
    tokensThatAreAlsoProps: ['vertical'],
    slotProps: {
        root: framework_1.buildProps(function (tokens) { return ({
            style: __assign(__assign({}, (tokens.vertical ? { borderLeftWidth: tokens.separatorWidth } : { borderTopWidth: tokens.separatorWidth })), (tokens.color && { borderColor: tokens.color })),
        }); }, ['color', 'vertical', 'separatorWidth']),
    },
    slots: { root: react_native_1.View },
    render: function (props, useSlots) {
        var Root = useSlots(props).root;
        return function (rest, children) { return framework_1.withSlots(Root, __assign({}, framework_1.mergeProps(props, rest, propMask)), children); };
    },
});
exports.default = exports.Separator;
