// @fluentui-react-native/use-tokens/lib-commonjs/applyPropsToTokens.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var applyPropsToTokens_1 = require("./applyPropsToTokens");
var memo_cache_1 = require("@fluentui-react-native/memo-cache");
var tokenProps = ['dos', 'quatro', 'cinco'];
var themeTokens = {
    uno: 'uno',
    dos: 'dos',
    tres: 3,
    quatro: 4,
    cinco: true,
};
var props1 = { dos: 'two', quatro: 'four', cinco: false, foo: 'foo', bar: 'bar' };
// const props2: Props = { dos: 'two' };
// const props3: Props = { foo: 'foo', bar: 'bar' };
describe('applyPropsToTokens tests', function () {
    test('props get copied', function () {
        var cache = memo_cache_1.getMemoCache();
        var tokens = applyPropsToTokens_1.applyPropsToTokens(props1, themeTokens, cache, tokenProps)[0];
        expect(tokens).not.toBe(themeTokens);
        for (var _i = 0, tokenProps_1 = tokenProps; _i < tokenProps_1.length; _i++) {
            var key = tokenProps_1[_i];
            expect(tokens[key]).toEqual(props1[key]);
        }
    });
    test('no copied props does not change tokens', function () {
        var cache = memo_cache_1.getMemoCache();
        var tokens = applyPropsToTokens_1.applyPropsToTokens({}, themeTokens, cache, tokenProps)[0];
        expect(tokens).toBe(themeTokens);
    });
});
