// @fluentui-react-native/tokens/lib-commonjs/text-tokens.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var foundation_tokens_1 = require("@uifabricshared/foundation-tokens");
exports.fontStyles = {
    from: function (_a, _b) {
        var fontFamily = _a.fontFamily, fontSize = _a.fontSize, fontWeight = _a.fontWeight, variant = _a.variant;
        var typography = _b.typography;
        var families = typography.families, sizes = typography.sizes, weights = typography.weights, variants = typography.variants;
        if (fontFamily || fontSize || fontWeight || variant) {
            return {
                fontFamily: families[fontFamily] || fontFamily || families[variants[variant].face] || variants[variant].face,
                fontSize: sizes[fontSize] || fontSize || sizes[variants[variant].size] || variants[variant].size,
                fontWeight: weights[fontWeight] || fontWeight || weights[variants[variant].weight] || variants[variant].weight,
            };
        }
        return {};
    },
    keys: ['fontFamily', 'fontSize', 'fontWeight', 'variant'],
};
function _buildTextStyles(tokens, theme) {
    return {
        style: exports.fontStyles.from(tokens, theme),
    };
}
exports.textTokens = foundation_tokens_1.styleFunction(_buildTextStyles, exports.fontStyles.keys);
