// @fluentui-react-native/button/lib/PrimaryButton/PrimaryButton.settings.win32.js
export var settings = [
    {
        tokens: {
            backgroundColor: 'brandBackground',
            color: 'neutralForegroundOnBrand',
            borderColor: 'brandBackground',
        },
        // TODO - #728: neutralForegroundOnBrand is not working for icon color.
        endIcon: {
            color: '#ffffff',
        },
        startIcon: {
            color: '#ffffff',
        },
        _overrides: {
            disabled: {
                tokens: {
                    backgroundColor: 'neutralBackgroundDisabled',
                    color: 'neutralForegroundDisabled',
                    borderColor: 'neutralBackgroundDisabled',
                },
            },
            hovered: {
                tokens: {
                    backgroundColor: 'brandBackgroundHover',
                    color: 'neutralForegroundOnBrand',
                    borderColor: 'brandBackgroundHover',
                },
            },
            pressed: {
                tokens: {
                    backgroundColor: 'brandBackgroundPressed',
                    color: 'neutralForegroundOnBrand',
                    borderColor: 'brandBackgroundPressed',
                },
            },
            focused: {
                tokens: {
                    backgroundColor: 'brandBackgroundHover',
                    borderColor: 'strokeFocus2',
                    color: 'neutralForegroundInvertedLinkHover',
                },
            },
        },
    },
    'PrimaryButton',
];
