// @fluentui-react-native/contextual-menu/lib/ContextualMenu.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
/** @jsx withSlots */
import * as React from 'react';
import { View } from 'react-native';
import { contextualMenuName, } from './ContextualMenu.types';
import { settings } from './ContextualMenu.settings';
import { compose } from '@uifabricshared/foundation-compose';
import { useSelectedKey } from '@fluentui-react-native/interactive-hooks';
import { mergeSettings } from '@uifabricshared/foundation-settings';
import { backgroundColorTokens, borderTokens } from '@fluentui-react-native/tokens';
import { Callout } from '@fluentui-react-native/callout';
import { withSlots } from '@uifabricshared/foundation-composable';
export var CMContext = React.createContext({
    selectedKey: null,
    onItemClick: function ( /* key: string */) {
        return;
    },
    onDismissMenu: function () {
        return;
    },
});
export var ContextualMenu = compose({
    displayName: contextualMenuName,
    usePrepareProps: function (userProps, useStyling) {
        var setShowMenu = userProps.setShowMenu, _a = userProps.shouldFocusOnMount, shouldFocusOnMount = _a === void 0 ? true : _a, _b = userProps.shouldFocusOnContainer, shouldFocusOnContainer = _b === void 0 ? false : _b, rest = __rest(userProps, ["setShowMenu", "shouldFocusOnMount", "shouldFocusOnContainer"]);
        // This hook updates the Selected Button and calls the customer's onClick function. This gets called after a button is pressed.
        var data = useSelectedKey(null, userProps.onItemClick);
        var dismissCallback = React.useCallback(function () {
            userProps.onDismiss();
            setShowMenu(false);
        }, [setShowMenu, userProps.onDismiss]);
        var _c = React.useState(true), containerFocus = _c[0], setContainerFocus = _c[1];
        var toggleContainerFocus = React.useCallback(function () {
            setContainerFocus(false);
        }, [setContainerFocus]);
        var state = {
            context: {
                selectedKey: data.selectedKey,
                onItemClick: data.onKeySelect,
                onDismissMenu: dismissCallback,
            },
        };
        var styleProps = useStyling(userProps, function (override) { return state[override] || userProps[override]; });
        var slotProps = mergeSettings(styleProps, {
            root: __assign(__assign({}, rest), { setInitialFocus: shouldFocusOnMount }),
            container: {
                accessible: shouldFocusOnContainer,
                focusable: shouldFocusOnContainer && containerFocus,
                onBlur: toggleContainerFocus,
            },
        });
        return { slotProps: slotProps, state: state };
    },
    settings: settings,
    slots: {
        root: Callout,
        container: View,
    },
    styles: {
        root: [backgroundColorTokens, borderTokens],
        container: [],
    },
    render: function (Slots, renderData) {
        var children = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            children[_i - 2] = arguments[_i];
        }
        if (renderData.state == undefined) {
            return null;
        }
        return (withSlots(CMContext.Provider, { value: renderData.state.context },
            withSlots(Slots.root, null,
                withSlots(Slots.container, null, children))));
    },
});
export default ContextualMenu;
