// @fluentui-react-native/contextual-menu/lib/SubmenuItem.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
/** @jsx withSlots */
import * as React from 'react';
import { View } from 'react-native';
import { submenuItemName, } from './SubmenuItem.types';
import { compose } from '@uifabricshared/foundation-compose';
import { withSlots } from '@uifabricshared/foundation-composable';
import { Text } from '@fluentui-react-native/text';
import { settings } from './SubmenuItem.settings';
import { backgroundColorTokens, borderTokens, textTokens, foregroundColorTokens, getPaletteFromTheme } from '@fluentui-react-native/tokens';
import { mergeSettings } from '@uifabricshared/foundation-settings';
import { useAsPressable, useKeyCallback, useViewCommandFocus } from '@fluentui-react-native/interactive-hooks';
import { SvgXml } from 'react-native-svg';
import { CMContext } from './ContextualMenu';
import { Icon } from '@fluentui-react-native/icon';
import { createIconProps } from '@fluentui-react-native/interactive-hooks';
export var SubmenuItem = compose({
    displayName: submenuItemName,
    usePrepareProps: function (userProps, useStyling) {
        var defaultComponentRef = React.useRef(null);
        var disabled = userProps.disabled, itemKey = userProps.itemKey, icon = userProps.icon, text = userProps.text, _a = userProps.accessibilityLabel, accessibilityLabel = _a === void 0 ? userProps.text : _a, onClick = userProps.onClick, testID = userProps.testID, _b = userProps.componentRef, componentRef = _b === void 0 ? defaultComponentRef : _b, rest = __rest(userProps, ["disabled", "itemKey", "icon", "text", "accessibilityLabel", "onClick", "testID", "componentRef"]);
        // Grabs the context information from Submenu (currently selected menuItem and client's onItemClick callback)
        var context = React.useContext(CMContext);
        var cmRef = useViewCommandFocus(componentRef);
        var onItemHoverIn = React.useCallback(function (e) {
            userProps.onHoverIn(e);
        }, []);
        var onItemClick = React.useCallback(function (e) {
            if (!disabled) {
                context === null || context === void 0 ? void 0 : context.onDismissMenu();
                (context === null || context === void 0 ? void 0 : context.dismissSubmenu) && context.dismissSubmenu();
                onClick ? onClick() : context === null || context === void 0 ? void 0 : context.onItemClick(itemKey);
                e.stopPropagation();
            }
        }, [context, disabled, itemKey, onClick]);
        var pressable = useAsPressable(__assign(__assign({}, rest), { onPress: onItemClick, onHoverIn: onItemHoverIn, delayHoverIn: 500 }));
        var _c = React.useState(false), submenuItemHovered = _c[0], setSubmenuItemHovered = _c[1];
        context.setSubmenuItemHovered = setSubmenuItemHovered;
        // set up state
        var state = __assign(__assign({}, pressable.state), { selected: context.selectedKey === userProps.itemKey, disabled: userProps.disabled, content: !!text, icon: !!icon, submenuItemHovered: submenuItemHovered });
        var onMouseEnter = React.useCallback(function (e) {
            setSubmenuItemHovered(true);
            pressable.props.onMouseEnter && pressable.props.onMouseEnter(e);
            e.stopPropagation();
        }, [pressable, setSubmenuItemHovered]);
        /*
         * SubmenuItem launches the submenu onMouseEnter event. For keyboarding, submenu should be launched with Spacebar, Enter, or right arrow.
         */
        var onKeyUp = useKeyCallback(onMouseEnter, ' ', 'Enter', 'ArrowRight');
        // grab the styling information, referencing the state as well as the props
        var styleProps = useStyling(userProps, function (override) { return state[override] || userProps[override]; });
        // create the merged slot props
        var slotProps = mergeSettings(styleProps, {
            root: __assign(__assign({}, pressable.props), { ref: cmRef, onKeyUp: onKeyUp,
                onMouseEnter: onMouseEnter,
                accessibilityLabel: accessibilityLabel }),
            content: { children: text, testID: testID },
            icon: createIconProps(icon),
            chevron: {},
        });
        return { slotProps: slotProps, state: state };
    },
    settings: settings,
    render: function (Slots, renderData) {
        var children = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            children[_i - 2] = arguments[_i];
        }
        var xml = "\n    <svg width=\"12\" height=\"12\" viewBox=\"0 0 2048 2048\" >\n      <path class='OfficeIconColors_HighContrast' fill='currentColor' d='M 743 1767 l -121 -121 l 708 -707 l -708 -708 l 121 -121 l 828 829 z' />\n      <path class='OfficeIconColors_m22' fill='currentColor' d='M 743 1767 l -121 -121 l 708 -707 l -708 -708 l 121 -121 l 828 829 z' />\n    </svg>";
        // We shouldn't have to specify the source prop on Slots.icon, here, but we need another drop from @uifabricshared
        return (withSlots(Slots.root, null,
            withSlots(Slots.leftstack, null,
                renderData.state.icon && withSlots(Slots.icon, null),
                renderData.state.content && withSlots(Slots.content, null),
                children),
            withSlots(Slots.rightstack, null,
                withSlots(Slots.chevron, { xml: xml }))));
    },
    slots: {
        root: View,
        leftstack: { slotType: View },
        icon: { slotType: Icon },
        content: Text,
        rightstack: { slotType: View },
        chevron: SvgXml,
    },
    styles: {
        root: [backgroundColorTokens, borderTokens],
        leftstack: [],
        icon: [{ source: 'iconColor', lookup: getPaletteFromTheme, target: 'color' }],
        content: [textTokens, foregroundColorTokens],
        rightstack: [],
        chevron: [{ source: 'chevronColor', lookup: getPaletteFromTheme, target: 'color' }],
    },
});
export default SubmenuItem;
