// @fluentui-react-native/persona-coin/lib/PersonaCoin.tokens.initialsBackground.js
import { styleFunction } from '@uifabricshared/foundation-tokens';
import { convertCoinColor, convertCoinColorFluent, calculateEffectiveSizes } from './PersonaCoin.helpers';
var _initialsBackgroundKeyProps = ['coinSize', 'size', 'coinColor', 'coinColorFluent', 'backgroundColor'];
function _buildInitialsBackgroundStyles(tokenProps /*, theme: ITheme*/) {
    var physicalSize = calculateEffectiveSizes(tokenProps).physicalSize;
    var backgroundColor = tokenProps.backgroundColor, coinColor = tokenProps.coinColor, coinColorFluent = tokenProps.coinColorFluent;
    var effectiveBackgroundColor = backgroundColor;
    if (coinColorFluent) {
        effectiveBackgroundColor = convertCoinColorFluent(coinColorFluent);
    }
    else if (coinColor) {
        effectiveBackgroundColor = convertCoinColor(coinColor);
    }
    return {
        style: {
            borderRadius: physicalSize / 2,
            width: physicalSize,
            height: physicalSize,
            flexGrow: 1,
            alignSelf: 'stretch',
            justifyContent: 'center',
            alignItems: 'center',
            backgroundColor: effectiveBackgroundColor,
        },
    };
}
export var buildInitialsBackgroundStyles = styleFunction(_buildInitialsBackgroundStyles, _initialsBackgroundKeyProps);
