// @fluentui-react-native/persona-coin/lib/PersonaCoin.tokens.ring.js
import { styleFunction } from '@uifabricshared/foundation-tokens';
import { getRingThickness, calculateEffectiveSizes } from './PersonaCoin.helpers';
var _ringProps = ['coinSize', 'size', 'ring'];
function _buildRingStyles(tokenProps, theme) {
    var physicalSize = calculateEffectiveSizes(tokenProps).physicalSize;
    var ring = tokenProps.ring;
    if (!ring)
        return {};
    var innerGap = ring.innerGap != undefined ? getRingThickness(ring.innerGap) : getRingThickness(ring.ringThickness || 'xxlarge');
    var effectiveRingThickness = 2 * getRingThickness(ring.ringThickness || 'xxlarge') + innerGap;
    var effectiveSize = physicalSize + 2 * effectiveRingThickness;
    var ringColor = ring.ringBackgroundColor || theme.colors.personaActivityRing;
    return {
        style: {
            borderStyle: 'solid',
            borderColor: ringColor,
            borderWidth: effectiveRingThickness,
            borderRadius: effectiveSize / 2,
            width: effectiveSize,
            height: effectiveSize,
            position: 'absolute',
            top: -effectiveRingThickness,
            left: -effectiveRingThickness,
        },
    };
}
function _buildGlowStyles(tokenProps, theme) {
    var physicalSize = calculateEffectiveSizes(tokenProps).physicalSize;
    var ring = tokenProps.ring;
    if (!ring)
        return {};
    var innerGap = ring.innerGap != undefined ? getRingThickness(ring.innerGap) : getRingThickness(ring.ringThickness || 'xxlarge');
    var effectiveRingThickness = getRingThickness(ring.ringThickness || 'xxlarge') + innerGap;
    var effectiveSize = physicalSize + 2 * effectiveRingThickness;
    var glowColor = ring.accent ? theme.colors.accentButtonBackground : ring.ringColor || theme.colors.personaActivityGlow;
    return {
        style: {
            borderStyle: 'solid',
            borderColor: glowColor,
            borderWidth: effectiveRingThickness - innerGap,
            borderRadius: effectiveSize / 2,
            width: effectiveSize,
            height: effectiveSize,
            position: 'absolute',
            top: -effectiveRingThickness,
            left: -effectiveRingThickness,
        },
    };
}
export var buildRingStyles = styleFunction(_buildRingStyles, _ringProps);
export var buildGlowStyles = styleFunction(_buildGlowStyles, _ringProps);
