// @fluentui-react-native/link/lib-commonjs/Link.settings.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var Link_types_1 = require("./Link.types");
exports.settings = [
    {
        tokens: {
            variant: 'secondaryStandard',
            color: 'link',
            borderColor: 'transparent',
            borderStyle: 'solid',
            borderWidth: 2,
            borderRadius: 4,
        },
        root: __assign(__assign({ accessible: true, focusable: true }, { cursor: 'pointer' }), { accessibilityRole: 'link', style: {
                display: 'flex',
                alignItems: 'flex-start',
            } }),
        content: {
            style: {
                textDecorationLine: 'underline',
                textAlign: 'center',
            },
        },
        _precedence: ['visited', 'hovered', 'focused', 'pressed', 'disabled'],
        _overrides: {
            disabled: {
                tokens: {
                    color: 'link',
                },
            },
            hovered: {
                tokens: {
                    color: 'linkHovered',
                },
            },
            pressed: {
                tokens: {
                    color: 'linkPressed',
                },
            },
            visited: {
                tokens: {
                    color: 'link',
                },
            },
            focused: {
                tokens: {
                    borderColor: 'strokeFocus2',
                },
            },
        },
    },
    Link_types_1.linkName,
];
