// @uifabricshared/foundation-tokens/lib/MockButton.js
import { standardTextTokens, standardForegroundColorTokens, standardBackgroundColorTokens, standardBorderTokens, standardCaptionTokens, } from './MockTokens';
import { mockCreate, stockFakeComponent } from './MockComponent';
/**
 * A mock simple text component with one slot that represents an internal text element.  This
 * supports text and fg color tokens.  These tokens should be delegated here and styles should be
 * produced by the MockText component itself for those tokens
 */
export var MockText = mockCreate({
    slots: {
        root: stockFakeComponent,
    },
    styles: {
        root: [standardTextTokens, standardForegroundColorTokens],
    },
});
/**
 * A mock button component that has a stock component for its root node, a sub-component for
 * text, and a stock component for the icon
 */
export var MockButton = mockCreate({
    slots: {
        root: stockFakeComponent,
        content: MockText,
        subContent: stockFakeComponent,
        icon: stockFakeComponent,
    },
    styles: {
        root: [standardBackgroundColorTokens, standardBorderTokens],
        content: [standardTextTokens, standardForegroundColorTokens],
        subContent: [standardTextTokens, standardCaptionTokens],
        icon: [standardForegroundColorTokens],
    },
});
