// @fluentui-react-native/default-theme/lib/defaultTheme.js
import { Platform } from 'react-native';
import { getStockWebPalette, getStockWebDarkPalette, getStockWebHCPalette } from './defaultColors';
import { globalTokens } from '@fluentui-react-native/theme-tokens';
function _defaultTypography() {
    var defaultsDict = {
        sizes: {
            caption: globalTokens.font.size['100'],
            secondary: globalTokens.font.size['200'],
            body: globalTokens.font.size['300'],
            subheader: globalTokens.font.size['400'],
            header: globalTokens.font.size['500'],
            hero: globalTokens.font.size['700'],
            heroLarge: 42,
        },
        weights: {
            regular: globalTokens.font.weight.regular,
            semiBold: globalTokens.font.weight.semibold,
        },
        families: {
            primary: 'Segoe UI',
            secondary: 'Segoe UI',
            cursive: 'System',
            monospace: 'System',
            sansSerif: 'System',
            serif: 'System',
        },
        variants: {
            captionStandard: { face: 'primary', size: 'caption', weight: 'regular' },
            secondaryStandard: { face: 'primary', size: 'secondary', weight: 'regular' },
            secondarySemibold: { face: 'primary', size: 'secondary', weight: 'semiBold' },
            bodyStandard: { face: 'primary', size: 'body', weight: 'regular' },
            bodySemibold: { face: 'primary', size: 'body', weight: 'semiBold' },
            subheaderStandard: { face: 'primary', size: 'subheader', weight: 'regular' },
            subheaderSemibold: { face: 'primary', size: 'subheader', weight: 'semiBold' },
            headerStandard: { face: 'primary', size: 'header', weight: 'regular' },
            headerSemibold: { face: 'primary', size: 'header', weight: 'semiBold' },
            heroStandard: { face: 'primary', size: 'hero', weight: 'regular' },
            heroSemibold: { face: 'primary', size: 'hero', weight: 'semiBold' },
            heroLargeStandard: { face: 'primary', size: 'heroLarge', weight: 'regular' },
            heroLargeSemibold: { face: 'primary', size: 'heroLarge', weight: 'semiBold' },
        },
    };
    if (Platform.OS === 'macos' || Platform.OS === 'ios') {
        var familiesDictApple = {
            primary: 'System',
            secondary: 'System',
            cursive: 'System',
            monospace: 'System',
            sansSerif: 'System',
            serif: 'System',
        };
        defaultsDict.families = familiesDictApple;
    }
    // In certain cases on win32, semibold font renders incorrectly so 'Segoe UI Semibold' is used
    if (Platform.OS === 'win32') {
        var familiesDictWin32 = {
            primary: 'Segoe UI',
            secondary: 'Segoe UI Semibold',
            cursive: 'System',
            monospace: 'System',
            sansSerif: 'System',
            serif: 'System',
        };
        defaultsDict.families = familiesDictWin32;
        var variantsDictWin32 = {
            captionStandard: { face: 'primary', size: 'caption', weight: 'regular' },
            secondaryStandard: { face: 'primary', size: 'secondary', weight: 'regular' },
            secondarySemibold: { face: 'secondary', size: 'secondary', weight: 'semiBold' },
            bodyStandard: { face: 'primary', size: 'body', weight: 'regular' },
            bodySemibold: { face: 'secondary', size: 'body', weight: 'semiBold' },
            subheaderStandard: { face: 'primary', size: 'subheader', weight: 'regular' },
            subheaderSemibold: { face: 'secondary', size: 'subheader', weight: 'semiBold' },
            headerStandard: { face: 'primary', size: 'header', weight: 'regular' },
            headerSemibold: { face: 'secondary', size: 'header', weight: 'semiBold' },
            heroStandard: { face: 'primary', size: 'hero', weight: 'regular' },
            heroSemibold: { face: 'secondary', size: 'hero', weight: 'semiBold' },
            heroLargeStandard: { face: 'primary', size: 'heroLarge', weight: 'regular' },
            heroLargeSemibold: { face: 'secondary', size: 'heroLarge', weight: 'semiBold' },
        };
        defaultsDict.variants = variantsDictWin32;
    }
    return defaultsDict;
}
export function defaultSpacing() {
    return { s2: '4px', s1: '8px', m: '16px', l1: '20px', l2: '32px' };
}
export var defaultFluentTheme = {
    colors: getStockWebPalette(),
    typography: _defaultTypography(),
    spacing: defaultSpacing(),
    components: {},
    host: { appearance: 'light' },
};
export var defaultFluentDarkTheme = {
    colors: getStockWebDarkPalette(),
    typography: defaultFluentTheme.typography,
    spacing: defaultFluentTheme.spacing,
    components: {},
    host: { appearance: 'dark' },
};
export var defaultFluentHighConstrastTheme = {
    colors: getStockWebHCPalette(),
    typography: defaultFluentTheme.typography,
    spacing: defaultFluentTheme.spacing,
    components: {},
    host: { appearance: 'highContrast' },
};
