// @fluentui-react-native/link/lib/Link.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
/** @jsx withSlots */
import * as React from 'react';
import { Linking, View } from 'react-native';
import { Text } from '@fluentui-react-native/text';
import { compose } from '@uifabricshared/foundation-compose';
import { linkName } from './Link.types';
import { settings } from './Link.settings';
import { foregroundColorTokens, textTokens, borderTokens } from '@fluentui-react-native/tokens';
import { useAsPressable, useKeyCallback, useOnPressWithFocus, useViewCommandFocus } from '@fluentui-react-native/interactive-hooks';
import { mergeSettings } from '@uifabricshared/foundation-settings';
import { withSlots } from '@uifabricshared/foundation-composable';
export function useAsLink(userProps, ref) {
    var url = userProps.url, onPress = userProps.onPress, rest = __rest(userProps, ["url", "onPress"]);
    var _a = React.useState({ visited: false }), linkState = _a[0], setLinkState = _a[1];
    var linkOnPress = React.useCallback(function (e) {
        setLinkState({ visited: true });
        if (url) {
            Linking.openURL(url);
        }
        else if (onPress) {
            onPress(e);
        }
    }, [setLinkState, url, onPress]);
    // Ensure focus is placed on link after click
    var linkOnPressWithFocus = useOnPressWithFocus(ref, linkOnPress);
    var pressable = useAsPressable(__assign({ onPress: linkOnPressWithFocus }, rest));
    var onKeyUp = useKeyCallback(linkOnPress, ' ', 'Enter');
    var newState = __assign(__assign({}, pressable.state), linkState);
    var newProps = __assign(__assign(__assign({}, userProps), pressable.props), { onKeyUp: onKeyUp });
    return [newProps, newState];
}
export var Link = compose({
    displayName: linkName,
    settings: settings,
    usePrepareProps: function (userProps, useStyling) {
        var defaultComponentRef = React.useRef(null);
        var content = userProps.content, onAccessibilityTap = userProps.onAccessibilityTap, _a = userProps.componentRef, componentRef = _a === void 0 ? defaultComponentRef : _a, rest = __rest(userProps, ["content", "onAccessibilityTap", "componentRef"]);
        var _b = useAsLink(rest, componentRef), linkProps = _b[0], linkState = _b[1];
        var onAccTap = onAccessibilityTap ? onAccessibilityTap : linkProps.onPress;
        var info = { content: !!content };
        var linkRef = useViewCommandFocus(componentRef);
        // grab the styling information, referencing the state as well as the props
        var styleProps = useStyling(userProps, function (override) { return linkState[override] || userProps[override]; });
        // create the merged slot props
        var slotProps = mergeSettings(styleProps, {
            root: __assign(__assign({}, linkProps), { ref: linkRef, onAccessibilityTap: onAccTap }),
            content: { children: content },
        });
        return { slotProps: slotProps, state: __assign(__assign({}, linkState), info) };
    },
    render: function (Slots, renderData) {
        var children = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            children[_i - 2] = arguments[_i];
        }
        var content = renderData.state && renderData.state.content;
        return children && children.length && children.length === 1 && children[0] !== undefined ? (withSlots(Slots.root, null,
            content && withSlots(Slots.content, null),
            children)) : (withSlots(Slots.root, null, content && withSlots(Slots.content, null)));
    },
    slots: {
        root: View,
        content: Text,
    },
    styles: {
        root: [],
        content: [foregroundColorTokens, textTokens, borderTokens],
    },
});
export default Link;
