// @uifabricshared/foundation-tokens/lib-commonjs/MockButton.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var MockTokens_1 = require("./MockTokens");
var MockComponent_1 = require("./MockComponent");
/**
 * A mock simple text component with one slot that represents an internal text element.  This
 * supports text and fg color tokens.  These tokens should be delegated here and styles should be
 * produced by the MockText component itself for those tokens
 */
exports.MockText = MockComponent_1.mockCreate({
    slots: {
        root: MockComponent_1.stockFakeComponent,
    },
    styles: {
        root: [MockTokens_1.standardTextTokens, MockTokens_1.standardForegroundColorTokens],
    },
});
/**
 * A mock button component that has a stock component for its root node, a sub-component for
 * text, and a stock component for the icon
 */
exports.MockButton = MockComponent_1.mockCreate({
    slots: {
        root: MockComponent_1.stockFakeComponent,
        content: exports.MockText,
        subContent: MockComponent_1.stockFakeComponent,
        icon: MockComponent_1.stockFakeComponent,
    },
    styles: {
        root: [MockTokens_1.standardBackgroundColorTokens, MockTokens_1.standardBorderTokens],
        content: [MockTokens_1.standardTextTokens, MockTokens_1.standardForegroundColorTokens],
        subContent: [MockTokens_1.standardTextTokens, MockTokens_1.standardCaptionTokens],
        icon: [MockTokens_1.standardForegroundColorTokens],
    },
});
