// @fluentui-react-native/checkbox/lib-commonjs/Checkbox.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
/** @jsx withSlots */
var React = __importStar(require("react"));
var react_native_1 = require("react-native");
var text_1 = require("@fluentui-react-native/text");
var Checkbox_types_1 = require("./Checkbox.types");
var foundation_compose_1 = require("@uifabricshared/foundation-compose");
var foundation_composable_1 = require("@uifabricshared/foundation-composable");
var adapters_1 = require("@fluentui-react-native/adapters");
var Checkbox_settings_1 = require("./Checkbox.settings");
var foundation_settings_1 = require("@uifabricshared/foundation-settings");
var tokens_1 = require("@fluentui-react-native/tokens");
var interactive_hooks_1 = require("@fluentui-react-native/interactive-hooks");
var tokens_2 = require("@fluentui-react-native/tokens");
exports.Checkbox = foundation_compose_1.compose({
    displayName: Checkbox_types_1.checkboxName,
    usePrepareProps: function (userProps, useStyling) {
        var _a;
        var defaultComponentRef = React.useRef(null);
        var accessible = userProps.accessible, accessibilityLabel = userProps.accessibilityLabel, accessibilityRole = userProps.accessibilityRole, ariaLabel = userProps.ariaLabel, checked = userProps.checked, defaultChecked = userProps.defaultChecked, boxSide = userProps.boxSide, label = userProps.label, onChange = userProps.onChange, _b = userProps.componentRef, componentRef = _b === void 0 ? defaultComponentRef : _b, rest = __rest(userProps, ["accessible", "accessibilityLabel", "accessibilityRole", "ariaLabel", "checked", "defaultChecked", "boxSide", "label", "onChange", "componentRef"]);
        // Warns defaultChecked and checked being mutually exclusive.
        if (defaultChecked != undefined && checked != undefined) {
            console.warn('defaultChecked and checked are mutually exclusive to one another. Use one or the other.');
        }
        // Re-usable hook for toggle components.
        var _c = interactive_hooks_1.useAsToggle(defaultChecked, checked, onChange), isChecked = _c[0], toggleChecked = _c[1];
        // Ensure focus is placed on checkbox after click
        var toggleCheckedWithFocus = interactive_hooks_1.useOnPressWithFocus(componentRef, toggleChecked);
        var pressable = interactive_hooks_1.useAsPressable(__assign({ onPress: toggleCheckedWithFocus }, rest));
        var buttonRef = interactive_hooks_1.useViewCommandFocus(componentRef);
        // Handles the "Space" key toggling the Checkbox
        var onKeyUpSpace = interactive_hooks_1.useKeyCallback(toggleChecked, ' ');
        var state = __assign(__assign({}, pressable.state), { disabled: !!userProps.disabled, checked: isChecked, boxAtEnd: boxSide == undefined || boxSide == 'start' ? false : true });
        // Grab the styling information from the userProps, referencing the state as well as the props.
        var styleProps = useStyling(userProps, function (override) { return state[override] || userProps[override]; });
        // Used when creating accessibility properties in mergeSettings below
        var onAccessibilityAction = React.useCallback(function (event) {
            switch (event.nativeEvent.actionName) {
                case 'Toggle':
                    toggleChecked();
                    break;
            }
        }, [toggleChecked, userProps, state, pressable.props]);
        var slotProps = foundation_settings_1.mergeSettings(styleProps, {
            root: __assign(__assign({ rest: rest, ref: buttonRef }, pressable.props), { accessible: accessible !== null && accessible !== void 0 ? accessible : true, accessibilityRole: accessibilityRole !== null && accessibilityRole !== void 0 ? accessibilityRole : 'checkbox', accessibilityLabel: (_a = accessibilityLabel !== null && accessibilityLabel !== void 0 ? accessibilityLabel : ariaLabel) !== null && _a !== void 0 ? _a : label, accessibilityState: { disabled: state.disabled, checked: state.checked }, accessibilityActions: [{ name: 'Toggle', label: Checkbox_settings_1.checkboxSelectActionLabel }], focusable: !state.disabled, onAccessibilityAction: onAccessibilityAction, onKeyUp: onKeyUpSpace }),
            // Temporary checkmark until SVG functionality
            checkmark: { children: '✓' },
            content: { children: label },
        });
        return { slotProps: slotProps, state: state };
    },
    render: function (Slots, renderData) {
        var children = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            children[_i - 2] = arguments[_i];
        }
        return (foundation_composable_1.withSlots(Slots.root, null,
            (renderData === null || renderData === void 0 ? void 0 : renderData.state.boxAtEnd) && foundation_composable_1.withSlots(Slots.content, null),
            foundation_composable_1.withSlots(Slots.checkbox, null,
                foundation_composable_1.withSlots(Slots.checkmark, null)),
            !(renderData === null || renderData === void 0 ? void 0 : renderData.state.boxAtEnd) && foundation_composable_1.withSlots(Slots.content, null),
            children));
    },
    settings: Checkbox_settings_1.settings,
    slots: {
        root: react_native_1.View,
        checkbox: { slotType: react_native_1.View, filter: adapters_1.filterViewProps },
        checkmark: text_1.Text,
        content: text_1.Text,
    },
    styles: {
        root: [],
        checkbox: [
            tokens_2.backgroundColorTokens,
            tokens_1.borderTokens,
            [
                { source: 'checkboxBackgroundColor', lookup: tokens_1.getPaletteFromTheme, target: 'backgroundColor' },
                { source: 'checkboxBorderColor', lookup: tokens_1.getPaletteFromTheme, target: 'borderColor' },
            ],
        ],
        checkmark: [
            tokens_1.foregroundColorTokens,
            [
                { source: 'checkmarkColor', lookup: tokens_1.getPaletteFromTheme, target: 'color' },
                { source: 'checkmarkVisibility', target: 'opacity' },
            ],
        ],
        content: [tokens_1.foregroundColorTokens, tokens_1.textTokens, [{ source: 'textBorderColor', lookup: tokens_1.getPaletteFromTheme, target: 'borderColor' }]],
    },
});
exports.default = exports.Checkbox;
