// @fluentui-react-native/adapters/lib-commonjs/adapters.android.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _viewMask = {
    children: true,
    accessible: true,
    accessibilityLabel: true,
    accessibilityRole: true,
    accessibilityState: true,
    accessibilityValue: true,
    accessibilityHint: true,
    collapsable: true,
    hitSlop: true,
    onLayout: true,
    pointerEvents: true,
    removeClippedSubviews: true,
    style: true,
    testID: true,
    nativeID: true,
    needsOffscreenAlphaCompositing: true,
    onStartShouldSetResponder: true,
    onStartShouldSetResponderCapture: true,
    onMoveShouldSetResponder: true,
    onMoveShouldSetResponderCapture: true,
    onResponderEnd: true,
    onResponderGrant: true,
    onResponderMove: true,
    onResponderReject: true,
    onAccessibilityAction: true,
    onAccessibilityEscape: true,
    onAccessibilityTap: true,
    onMagicTap: true,
    onResponderRelease: true,
    onResponderStart: true,
    onResponderTerminate: true,
    onResponderTerminationRequest: true,
    onTouchCancel: true,
    onTouchEnd: true,
    onTouchEndCapture: true,
    onTouchMove: true,
    onTouchStart: true,
    renderToHardwareTextureAndroid: true,
    importantForAccessibility: true,
    accessibilityActions: true,
    accessibilityElementsHidden: true,
    accessibilityIgnoresInvertColors: true,
    accessibilityLiveRegion: true,
    accessibilityViewIsModal: true,
    focusable: true,
};
var _textMask = {
    children: true,
    accessibilityActions: true,
    accessibilityElementsHidden: true,
    accessibilityHint: true,
    accessibilityIgnoresInvertColors: true,
    accessibilityLabel: true,
    accessibilityLiveRegion: true,
    accessibilityRole: true,
    accessibilityState: true,
    accessibilityValue: true,
    accessibilityViewIsModal: true,
    accessible: true,
    onAccessibilityAction: true,
    onAccessibilityEscape: true,
    onAccessibilityTap: true,
    importantForAccessibility: true,
    testID: true,
    allowFontScaling: true,
    ellipsizeMode: true,
    lineBreakMode: true,
    numberOfLines: true,
    onLayout: true,
    onLongPress: true,
    onMagicTap: true,
    onPress: true,
    selectable: true,
    selectionColor: true,
    style: true,
    textBreakStrategy: true,
    nativeID: true,
    maxFontSizeMultiplier: true,
    onTextLayout: true,
    dataDetectorType: true,
};
var _imageMask = {
    children: true,
    accessibilityActions: true,
    accessibilityElementsHidden: true,
    accessibilityHint: true,
    accessibilityIgnoresInvertColors: true,
    accessibilityLabel: true,
    accessibilityLiveRegion: true,
    accessibilityRole: true,
    accessibilityState: true,
    accessibilityValue: true,
    accessibilityViewIsModal: true,
    accessible: true,
    onAccessibilityAction: true,
    onAccessibilityEscape: true,
    onAccessibilityTap: true,
    importantForAccessibility: true,
    style: true,
    onError: true,
    onLayout: true,
    onLoad: true,
    onLoadEnd: true,
    onLoadStart: true,
    onMagicTap: true,
    borderBottomLeftRadius: true,
    borderBottomRightRadius: true,
    borderRadius: true,
    borderTopLeftRadius: true,
    borderTopRightRadius: true,
    progressiveRenderingEnabled: true,
    source: true,
    resizeMethod: true,
    resizeMode: true,
    loadingIndicatorSource: true,
    testID: true,
    defaultSource: true,
    width: true,
    height: true,
    fadeDuration: true,
};
function filterViewProps(propName) {
    return _viewMask[propName];
}
exports.filterViewProps = filterViewProps;
function filterTextProps(propName) {
    return _textMask[propName];
}
exports.filterTextProps = filterTextProps;
function filterImageProps(propName) {
    return _imageMask[propName];
}
exports.filterImageProps = filterImageProps;
