// @fluentui-react-native/adapters/lib-commonjs/adapters.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _viewMask = {
    children: true,
    accessible: true,
    accessibilityControls: true,
    accessibilityLabel: true,
    accessibilityRole: true,
    accessibilityState: true,
    accessibilityHint: true,
    accessibilityValue: true,
    hitSlop: true,
    onLayout: true,
    pointerEvents: true,
    removeClippedSubviews: true,
    style: true,
    testID: true,
    nativeID: true,
    onStartShouldSetResponder: true,
    onStartShouldSetResponderCapture: true,
    onMoveShouldSetResponder: true,
    onMoveShouldSetResponderCapture: true,
    onResponderEnd: true,
    onResponderGrant: true,
    onResponderMove: true,
    onResponderReject: true,
    onAccessibilityAction: true,
    onAccessibilityEscape: true,
    onAccessibilityTap: true,
    onMagicTap: true,
    onResponderRelease: true,
    onResponderStart: true,
    onResponderTerminate: true,
    onResponderTerminationRequest: true,
    onTouchCancel: true,
    onTouchEnd: true,
    onTouchEndCapture: true,
    onTouchMove: true,
    onTouchStart: true,
    importantForAccessibility: true,
    accessibilityActions: true,
    accessibilityElementsHidden: true,
    accessibilityIgnoresInvertColors: true,
    accessibilityLiveRegion: true,
    accessibilityViewIsModal: true,
    collapsable: true,
    needsOffscreenAlphaCompositing: true,
    isTVSelectable: true,
    renderToHardwareTextureAndroid: true,
    shouldRasterizeIOS: true,
    hasTVPreferredFocus: true,
    tvParallaxMagnification: true,
    tvParallaxProperties: true,
    tvParallaxShiftDistanceX: true,
    tvParallaxShiftDistanceY: true,
    tvParallaxTiltAngle: true,
    focusable: true,
    accessibilityAnnotation: true,
    accessibilityDescription: true,
    accessibilityDescribedBy: true,
    accessibilityLabeledBy: true,
    accessibilityLevel: true,
    accessibilityPositionInSet: true,
    accessibilitySetSize: true,
    animationClass: true,
    cursor: true,
    keyDownEvents: true,
    keyUpEvents: true,
    onBlur: true,
    onBlurCapture: true,
    onFocus: true,
    onFocusCapture: true,
    onKeyDown: true,
    onKeyDownCapture: true,
    onKeyUp: true,
    onKeyUpCapture: true,
    onMouseEnter: true,
    onMouseLeave: true,
    tooltip: true,
    type: true,
};
var _textMask = {
    children: true,
    accessibilityActions: true,
    accessibilityControls: true,
    accessibilityElementsHidden: true,
    accessibilityHint: true,
    accessibilityIgnoresInvertColors: true,
    accessibilityLabel: true,
    accessibilityLiveRegion: true,
    accessibilityRole: true,
    accessibilityState: true,
    accessibilityValue: true,
    accessibilityViewIsModal: true,
    accessible: true,
    onAccessibilityAction: true,
    onAccessibilityEscape: true,
    onAccessibilityTap: true,
    importantForAccessibility: true,
    textBreakStrategy: true,
    testID: true,
    allowFontScaling: true,
    ellipsizeMode: true,
    lineBreakMode: true,
    numberOfLines: true,
    onLayout: true,
    onLongPress: true,
    onMagicTap: true,
    onPress: true,
    style: true,
    nativeID: true,
    maxFontSizeMultiplier: true,
    adjustsFontSizeToFit: true,
    minimumFontScale: true,
    selectable: true,
    selectionColor: true,
    suppressHighlighting: true,
    accessibilityDescribedBy: true,
    accessibilityLabeledBy: true,
    keyDownEvents: true,
    keyUpEvents: true,
    onKeyDown: true,
    onKeyDownCapture: true,
    onKeyUp: true,
    onKeyUpCapture: true,
    textStyle: true,
    tooltip: true,
    onTextLayout: true,
    dataDetectorType: true,
    focusable: true,
    onBlur: true,
    onBlurCapture: true,
    onFocus: true,
    onFocusCapture: true,
};
var _imageMask = {
    children: true,
    accessibilityActions: true,
    accessibilityElementsHidden: true,
    accessibilityHint: true,
    accessibilityIgnoresInvertColors: true,
    accessibilityLabel: true,
    accessibilityLiveRegion: true,
    accessibilityRole: true,
    accessibilityState: true,
    accessibilityValue: true,
    accessibilityViewIsModal: true,
    accessible: true,
    onAccessibilityAction: true,
    onAccessibilityEscape: true,
    onAccessibilityTap: true,
    importantForAccessibility: true,
    style: true,
    onError: true,
    onLayout: true,
    onLoad: true,
    onLoadEnd: true,
    onLoadStart: true,
    onMagicTap: true,
    onPartialLoad: true,
    onProgress: true,
    borderBottomLeftRadius: true,
    borderBottomRightRadius: true,
    borderRadius: true,
    borderTopLeftRadius: true,
    borderTopRightRadius: true,
    progressiveRenderingEnabled: true,
    source: true,
    resizeMethod: true,
    resizeMode: true,
    loadingIndicatorSource: true,
    testID: true,
    defaultSource: true,
    blurRadius: true,
    capInsets: true,
    width: true,
    height: true,
    fadeDuration: true,
};
function filterViewProps(propName) {
    return _viewMask[propName];
}
exports.filterViewProps = filterViewProps;
function filterTextProps(propName) {
    return _textMask[propName];
}
exports.filterTextProps = filterTextProps;
function filterImageProps(propName) {
    return _imageMask[propName];
}
exports.filterImageProps = filterImageProps;
