// @fluentui-react-native/tabs/lib-commonjs/Tabs.windows.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
/** @jsx withSlots */
var React = __importStar(require("react"));
var react_native_1 = require("react-native");
var text_1 = require("@fluentui-react-native/text");
var Tabs_types_1 = require("./Tabs.types");
var foundation_compose_1 = require("@uifabricshared/foundation-compose");
var foundation_composable_1 = require("@uifabricshared/foundation-composable");
var Tabs_settings_1 = require("./Tabs.settings");
var foundation_settings_1 = require("@uifabricshared/foundation-settings");
var adapters_1 = require("@fluentui-react-native/adapters");
var tokens_1 = require("@fluentui-react-native/tokens");
var interactive_hooks_1 = require("@fluentui-react-native/interactive-hooks");
exports.TabsContext = React.createContext({
    selectedKey: null,
    onTabsClick: function ( /* key: string */) {
        return;
    },
    getTabId: function ( /* key:string, index: number*/) {
        return null;
    },
    updateSelectedTabsItemRef: function ( /* ref: React.RefObject<any>*/) {
        return;
    },
    tabsItemKeys: [],
    views: null,
    focusZoneRef: null,
});
exports.Tabs = foundation_compose_1.compose({
    displayName: Tabs_types_1.tabsName,
    usePrepareProps: function (userProps, useStyling) {
        var focusZoneRef = React.useRef(null);
        var defaultComponentRef = React.useRef(null);
        var label = userProps.label, _a = userProps.accessibilityLabel, accessibilityLabel = _a === void 0 ? userProps.label : _a, selectedKey = userProps.selectedKey, headersOnly = userProps.headersOnly, defaultSelectedKey = userProps.defaultSelectedKey, getTabId = userProps.getTabId, _b = userProps.componentRef, componentRef = _b === void 0 ? defaultComponentRef : _b, isCircularNavigation = userProps.isCircularNavigation, rest = __rest(userProps, ["label", "accessibilityLabel", "selectedKey", "headersOnly", "defaultSelectedKey", "getTabId", "componentRef", "isCircularNavigation"]);
        // This hook updates the Selected TabsItem and calls the customer's onTabsClick function. This gets called after a TabsItem is pressed.
        var data = interactive_hooks_1.useSelectedKey(selectedKey || defaultSelectedKey || null, userProps.onTabsClick);
        var findTabId = React.useCallback(function (key, index) {
            if (getTabId) {
                return getTabId(key, index);
            }
            return key + "-Tab" + index;
        }, [getTabId]);
        // Stores views to be displayed
        var map = new Map();
        var state = {
            context: {
                selectedKey: selectedKey !== null && selectedKey !== void 0 ? selectedKey : data.selectedKey,
                onTabsClick: data.onKeySelect,
                getTabId: findTabId,
                views: map,
                focusZoneRef: focusZoneRef,
            },
            info: {
                headersOnly: headersOnly !== null && headersOnly !== void 0 ? headersOnly : false,
                label: !!label,
            },
        };
        var styleProps = useStyling(userProps, function (override) { return state[override] || userProps[override]; });
        var pressable = interactive_hooks_1.useAsPressable(__assign({}, rest));
        var onKeyDown = function (ev) {
            if (ev.nativeEvent.key === 'ArrowRight' || ev.nativeEvent.key === 'ArrowLeft') {
                var length_1 = state.info.enabledKeys.length;
                var currTabItemIndex = state.info.enabledKeys.findIndex(function (x) { return x == state.context.selectedKey; });
                var newCurrTabItemIndex = void 0;
                if (ev.nativeEvent.key === 'ArrowRight') {
                    if (isCircularNavigation || !(currTabItemIndex + 1 == length_1)) {
                        newCurrTabItemIndex = (currTabItemIndex + 1) % length_1;
                        state.context.selectedKey = state.info.enabledKeys[newCurrTabItemIndex];
                        data.onKeySelect(state.context.selectedKey);
                    }
                }
                else {
                    if (isCircularNavigation || !(currTabItemIndex == 0)) {
                        newCurrTabItemIndex = (currTabItemIndex - 1 + length_1) % length_1;
                        state.context.selectedKey = state.info.enabledKeys[newCurrTabItemIndex];
                        data.onKeySelect(state.context.selectedKey);
                    }
                }
            }
        };
        /* GH #964, Extra props are needed because FocusZone is not implemented on windows.
        The ref focusZoneRef is used to set focus on Tabs when selecting a TabsItem and onKeyDown manages keyboarding */
        var slotProps = foundation_settings_1.mergeSettings(styleProps, {
            root: __assign(__assign({ ref: componentRef, accessibilityLabel: accessibilityLabel, accessibilityRole: 'tablist' }, pressable.props), rest),
            label: { children: label },
            stack: { focusable: true, ref: focusZoneRef, onKeyDown: onKeyDown },
        });
        return { slotProps: slotProps, state: state };
    },
    render: function (Slots, renderData) {
        var children = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            children[_i - 2] = arguments[_i];
        }
        var _a, _b, _c, _d;
        if (!renderData.state) {
            return null;
        }
        // Populate the tabsItemKeys array
        if (children) {
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore - TODO, fix typing error
            renderData.state.context.tabsItemKeys = React.Children.map(children, function (child) {
                if (React.isValidElement(child)) {
                    return child.props.itemKey;
                }
            });
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore - TODO, fix typing error
            renderData.state.info.enabledKeys = React.Children.map(children, function (child) {
                if (React.isValidElement(child)) {
                    if (!child.props.disabled) {
                        return child.props.itemKey;
                    }
                }
            });
            /* Sets the default selected TabsItem if a TabsItem is hidden.
            The default selected Tabsitem is the first enabled TabsItem. */
            if (!renderData.state.info.enabledKeys.includes(renderData.state.context.selectedKey)) {
                renderData.state.context.selectedKey = (_a = renderData.state.info.enabledKeys[0]) !== null && _a !== void 0 ? _a : null;
            }
        }
        return (foundation_composable_1.withSlots(exports.TabsContext.Provider
        // Passes in the selected key and a hook function to update the newly selected tabsItem and call the client's onTabsClick callback
        , { 
            // Passes in the selected key and a hook function to update the newly selected tabsItem and call the client's onTabsClick callback
            value: (_b = renderData.state) === null || _b === void 0 ? void 0 : _b.context },
            foundation_composable_1.withSlots(Slots.root, null,
                ((_d = (_c = renderData.state) === null || _c === void 0 ? void 0 : _c.info) === null || _d === void 0 ? void 0 : _d.label) && foundation_composable_1.withSlots(Slots.label, null),
                foundation_composable_1.withSlots(Slots.stack, null, children),
                foundation_composable_1.withSlots(Slots.tabPanel, null,
                    foundation_composable_1.withSlots(exports.TabsContext.Consumer, null, function (context) { return !renderData.state.info.headersOnly && context.views.get(context.selectedKey); })))));
    },
    settings: Tabs_settings_1.settings,
    slots: {
        root: react_native_1.View,
        label: text_1.Text,
        stack: react_native_1.View,
        tabPanel: { slotType: react_native_1.View, filter: adapters_1.filterViewProps },
    },
    styles: {
        root: [],
        label: [tokens_1.foregroundColorTokens, tokens_1.textTokens],
        stack: [tokens_1.backgroundColorTokens],
    },
});
exports.default = exports.Tabs;
