// @fluentui-react-native/adapters/lib/adapters.ios.js
var _viewMask = {
    children: true,
    accessible: true,
    accessibilityLabel: true,
    accessibilityRole: true,
    accessibilityState: true,
    accessibilityHint: true,
    accessibilityValue: true,
    hitSlop: true,
    onLayout: true,
    pointerEvents: true,
    removeClippedSubviews: true,
    style: true,
    testID: true,
    nativeID: true,
    onStartShouldSetResponder: true,
    onStartShouldSetResponderCapture: true,
    onMoveShouldSetResponder: true,
    onMoveShouldSetResponderCapture: true,
    onResponderEnd: true,
    onResponderGrant: true,
    onResponderMove: true,
    onResponderReject: true,
    onAccessibilityAction: true,
    onAccessibilityEscape: true,
    onAccessibilityTap: true,
    onMagicTap: true,
    onResponderRelease: true,
    onResponderStart: true,
    onResponderTerminate: true,
    onResponderTerminationRequest: true,
    onTouchCancel: true,
    onTouchEnd: true,
    onTouchEndCapture: true,
    onTouchMove: true,
    onTouchStart: true,
    importantForAccessibility: true,
    accessibilityActions: true,
    accessibilityElementsHidden: true,
    accessibilityIgnoresInvertColors: true,
    accessibilityLiveRegion: true,
    accessibilityViewIsModal: true,
    isTVSelectable: true,
    shouldRasterizeIOS: true,
    hasTVPreferredFocus: true,
    tvParallaxMagnification: true,
    tvParallaxProperties: true,
    tvParallaxShiftDistanceX: true,
    tvParallaxShiftDistanceY: true,
    tvParallaxTiltAngle: true,
};
var _textMask = {
    children: true,
    accessibilityActions: true,
    accessibilityElementsHidden: true,
    accessibilityHint: true,
    accessibilityIgnoresInvertColors: true,
    accessibilityLabel: true,
    accessibilityLiveRegion: true,
    accessibilityRole: true,
    accessibilityState: true,
    accessibilityValue: true,
    accessibilityViewIsModal: true,
    accessible: true,
    onAccessibilityAction: true,
    onAccessibilityEscape: true,
    onAccessibilityTap: true,
    importantForAccessibility: true,
    testID: true,
    allowFontScaling: true,
    ellipsizeMode: true,
    lineBreakMode: true,
    numberOfLines: true,
    onLayout: true,
    onLongPress: true,
    onMagicTap: true,
    onPress: true,
    style: true,
    nativeID: true,
    maxFontSizeMultiplier: true,
    adjustsFontSizeToFit: true,
    minimumFontScale: true,
    suppressHighlighting: true,
    onTextLayout: true,
};
var _imageMask = {
    children: true,
    accessibilityActions: true,
    accessibilityElementsHidden: true,
    accessibilityHint: true,
    accessibilityIgnoresInvertColors: true,
    accessibilityLabel: true,
    accessibilityLiveRegion: true,
    accessibilityRole: true,
    accessibilityState: true,
    accessibilityValue: true,
    accessibilityViewIsModal: true,
    accessible: true,
    onAccessibilityAction: true,
    onAccessibilityEscape: true,
    onAccessibilityTap: true,
    importantForAccessibility: true,
    style: true,
    onError: true,
    onLayout: true,
    onLoad: true,
    onLoadEnd: true,
    onLoadStart: true,
    onMagicTap: true,
    onPartialLoad: true,
    onProgress: true,
    borderBottomLeftRadius: true,
    borderBottomRightRadius: true,
    borderRadius: true,
    borderTopLeftRadius: true,
    borderTopRightRadius: true,
    progressiveRenderingEnabled: true,
    source: true,
    resizeMode: true,
    loadingIndicatorSource: true,
    testID: true,
    defaultSource: true,
    blurRadius: true,
    capInsets: true,
};
export function filterViewProps(propName) {
    return _viewMask[propName];
}
export function filterTextProps(propName) {
    return _textMask[propName];
}
export function filterImageProps(propName) {
    return _imageMask[propName];
}
