// @fluentui-react-native/tabs/lib/Tabs.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
/** @jsx withSlots */
import * as React from 'react';
import { View } from 'react-native';
import { Text } from '@fluentui-react-native/text';
import { FocusZone } from '@fluentui-react-native/focus-zone';
import { tabsName } from './Tabs.types';
import { compose } from '@uifabricshared/foundation-compose';
import { withSlots } from '@uifabricshared/foundation-composable';
import { settings } from './Tabs.settings';
import { mergeSettings } from '@uifabricshared/foundation-settings';
import { filterViewProps } from '@fluentui-react-native/adapters';
import { foregroundColorTokens, textTokens, backgroundColorTokens } from '@fluentui-react-native/tokens';
import { useSelectedKey } from '@fluentui-react-native/interactive-hooks';
export var TabsContext = React.createContext({
    selectedKey: null,
    onTabsClick: function ( /* key: string */) {
        return;
    },
    getTabId: function ( /* key:string, index: number*/) {
        return null;
    },
    updateSelectedTabsItemRef: function ( /* ref: React.RefObject<any>*/) {
        return;
    },
    tabsItemKeys: [],
    views: null,
});
export var Tabs = compose({
    displayName: tabsName,
    usePrepareProps: function (userProps, useStyling) {
        var defaultComponentRef = React.useRef(null);
        var label = userProps.label, _a = userProps.accessibilityLabel, accessibilityLabel = _a === void 0 ? userProps.label : _a, selectedKey = userProps.selectedKey, headersOnly = userProps.headersOnly, defaultSelectedKey = userProps.defaultSelectedKey, getTabId = userProps.getTabId, _b = userProps.componentRef, componentRef = _b === void 0 ? defaultComponentRef : _b, isCircularNavigation = userProps.isCircularNavigation, rest = __rest(userProps, ["label", "accessibilityLabel", "selectedKey", "headersOnly", "defaultSelectedKey", "getTabId", "componentRef", "isCircularNavigation"]);
        /* The useSelectedKey hook is called after a TabsItem is pressed to update the Tab and TabsItem
        selected property and call the Tab and TabsItem onTabsClick callback. */
        var data = useSelectedKey(selectedKey || defaultSelectedKey || null, userProps.onTabsClick);
        var _c = React.useState(React.useRef(null)), selectedTabsItemRef = _c[0], setSelectedTabsItemRef = _c[1];
        var onSelectTabsItemRef = React.useCallback(function (ref) {
            setSelectedTabsItemRef(ref);
        }, [setSelectedTabsItemRef]);
        var findTabId = React.useCallback(function (key, index) {
            if (getTabId) {
                return getTabId(key, index);
            }
            return key + "-Tab" + index;
        }, [getTabId]);
        // Stores views to be displayed.
        var map = new Map();
        var state = {
            context: {
                selectedKey: selectedKey !== null && selectedKey !== void 0 ? selectedKey : data.selectedKey,
                onTabsClick: data.onKeySelect,
                getTabId: findTabId,
                updateSelectedTabsItemRef: onSelectTabsItemRef,
                views: map,
            },
            info: {
                headersOnly: headersOnly !== null && headersOnly !== void 0 ? headersOnly : false,
                label: !!label,
            },
        };
        var styleProps = useStyling(userProps, function (override) { return state[override] || userProps[override]; });
        var slotProps = mergeSettings(styleProps, {
            root: __assign({ ref: componentRef, accessibilityLabel: accessibilityLabel, accessibilityRole: 'tablist' }, rest),
            label: { children: label },
            container: { isCircularNavigation: isCircularNavigation, defaultTabbableElement: selectedTabsItemRef },
        });
        return { slotProps: slotProps, state: state };
    },
    render: function (Slots, renderData) {
        var children = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            children[_i - 2] = arguments[_i];
        }
        var _a, _b, _c, _d;
        if (!renderData.state) {
            return null;
        }
        // Populate the tabsItemKeys array.
        if (children) {
            var enabledKeys_1 = [];
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore - TODO, fix typing error
            // Generates array of keys and enabled keys.
            renderData.state.context.tabsItemKeys = React.Children.map(children, function (child) {
                if (React.isValidElement(child)) {
                    if (!child.props.disabled) {
                        enabledKeys_1.push(child.props.itemKey);
                    }
                    return child.props.itemKey;
                }
            });
            /* Sets the default selected TabsItem if a TabsItem is hidden.
            The default selected Tabsitem is the first enabled TabsItem. */
            if (!enabledKeys_1.includes(renderData.state.context.selectedKey)) {
                renderData.state.context.selectedKey = (_a = enabledKeys_1[0]) !== null && _a !== void 0 ? _a : null;
            }
        }
        return (withSlots(TabsContext.Provider
        // Passes in the selected key and a hook function to update the newly selected tabsItem and call the client's onTabsClick callback.
        , { 
            // Passes in the selected key and a hook function to update the newly selected tabsItem and call the client's onTabsClick callback.
            value: (_b = renderData.state) === null || _b === void 0 ? void 0 : _b.context },
            withSlots(Slots.root, null,
                ((_d = (_c = renderData.state) === null || _c === void 0 ? void 0 : _c.info) === null || _d === void 0 ? void 0 : _d.label) && withSlots(Slots.label, null),
                withSlots(Slots.container, null,
                    withSlots(Slots.stack, null, children)),
                withSlots(Slots.tabPanel, null,
                    withSlots(TabsContext.Consumer, null, function (context) { return !renderData.state.info.headersOnly && context.views.get(context.selectedKey); })))));
    },
    settings: settings,
    slots: {
        root: View,
        label: Text,
        container: FocusZone,
        stack: { slotType: View, filter: filterViewProps },
        tabPanel: { slotType: View, filter: filterViewProps },
    },
    styles: {
        root: [],
        label: [foregroundColorTokens, textTokens],
        stack: [backgroundColorTokens],
    },
});
export default Tabs;
