// @fluentui-react-native/persona-coin/lib-commonjs/PersonaCoin.tokens.ring.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var foundation_tokens_1 = require("@uifabricshared/foundation-tokens");
var PersonaCoin_helpers_1 = require("./PersonaCoin.helpers");
var _ringProps = ['coinSize', 'size', 'ring'];
function _buildRingStyles(tokenProps, theme) {
    var physicalSize = PersonaCoin_helpers_1.calculateEffectiveSizes(tokenProps).physicalSize;
    var ring = tokenProps.ring;
    if (!ring)
        return {};
    var innerGap = ring.innerGap != undefined ? PersonaCoin_helpers_1.getRingThickness(ring.innerGap) : PersonaCoin_helpers_1.getRingThickness(ring.ringThickness || 'xxlarge');
    var effectiveRingThickness = 2 * PersonaCoin_helpers_1.getRingThickness(ring.ringThickness || 'xxlarge') + innerGap;
    var effectiveSize = physicalSize + 2 * effectiveRingThickness;
    var ringColor = ring.ringBackgroundColor || theme.colors.personaActivityRing;
    return {
        style: {
            borderStyle: 'solid',
            borderColor: ringColor,
            borderWidth: effectiveRingThickness,
            borderRadius: effectiveSize / 2,
            width: effectiveSize,
            height: effectiveSize,
            position: 'absolute',
            top: -effectiveRingThickness,
            left: -effectiveRingThickness,
        },
    };
}
function _buildGlowStyles(tokenProps, theme) {
    var physicalSize = PersonaCoin_helpers_1.calculateEffectiveSizes(tokenProps).physicalSize;
    var ring = tokenProps.ring;
    if (!ring)
        return {};
    var innerGap = ring.innerGap != undefined ? PersonaCoin_helpers_1.getRingThickness(ring.innerGap) : PersonaCoin_helpers_1.getRingThickness(ring.ringThickness || 'xxlarge');
    var effectiveRingThickness = PersonaCoin_helpers_1.getRingThickness(ring.ringThickness || 'xxlarge') + innerGap;
    var effectiveSize = physicalSize + 2 * effectiveRingThickness;
    var glowColor = ring.accent ? theme.colors.accentButtonBackground : ring.ringColor || theme.colors.personaActivityGlow;
    return {
        style: {
            borderStyle: 'solid',
            borderColor: glowColor,
            borderWidth: effectiveRingThickness - innerGap,
            borderRadius: effectiveSize / 2,
            width: effectiveSize,
            height: effectiveSize,
            position: 'absolute',
            top: -effectiveRingThickness,
            left: -effectiveRingThickness,
        },
    };
}
exports.buildRingStyles = foundation_tokens_1.styleFunction(_buildRingStyles, _ringProps);
exports.buildGlowStyles = foundation_tokens_1.styleFunction(_buildGlowStyles, _ringProps);
