// @fluentui-react-native/persona-coin/lib-commonjs/PersonaCoin.tokens.initialsBackground.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var foundation_tokens_1 = require("@uifabricshared/foundation-tokens");
var PersonaCoin_helpers_1 = require("./PersonaCoin.helpers");
var _initialsBackgroundKeyProps = ['coinSize', 'size', 'coinColor', 'coinColorFluent', 'backgroundColor'];
function _buildInitialsBackgroundStyles(tokenProps /*, theme: ITheme*/) {
    var physicalSize = PersonaCoin_helpers_1.calculateEffectiveSizes(tokenProps).physicalSize;
    var backgroundColor = tokenProps.backgroundColor, coinColor = tokenProps.coinColor, coinColorFluent = tokenProps.coinColorFluent;
    var effectiveBackgroundColor = backgroundColor;
    if (coinColorFluent) {
        effectiveBackgroundColor = PersonaCoin_helpers_1.convertCoinColorFluent(coinColorFluent);
    }
    else if (coinColor) {
        effectiveBackgroundColor = PersonaCoin_helpers_1.convertCoinColor(coinColor);
    }
    return {
        style: {
            borderRadius: physicalSize / 2,
            width: physicalSize,
            height: physicalSize,
            flexGrow: 1,
            alignSelf: 'stretch',
            justifyContent: 'center',
            alignItems: 'center',
            backgroundColor: effectiveBackgroundColor,
        },
    };
}
exports.buildInitialsBackgroundStyles = foundation_tokens_1.styleFunction(_buildInitialsBackgroundStyles, _initialsBackgroundKeyProps);
