// @fluentui-react-native/radio-group/lib-commonjs/RadioButton.js
/** @jsx withSlots */
'use strict';
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = __importStar(require("react"));
var react_native_1 = require("react-native");
var text_1 = require("@fluentui-react-native/text");
var RadioButton_types_1 = require("./RadioButton.types");
var foundation_compose_1 = require("@uifabricshared/foundation-compose");
var adapters_1 = require("@fluentui-react-native/adapters");
var foundation_composable_1 = require("@uifabricshared/foundation-composable");
var RadioButton_settings_1 = require("./RadioButton.settings");
var foundation_settings_1 = require("@uifabricshared/foundation-settings");
var tokens_1 = require("@fluentui-react-native/tokens");
var interactive_hooks_1 = require("@fluentui-react-native/interactive-hooks");
var RadioGroup_1 = require("./RadioGroup");
exports.RadioButton = foundation_compose_1.compose({
    displayName: RadioButton_types_1.radioButtonName,
    usePrepareProps: function (userProps, useStyling) {
        var _a, _b, _c;
        var defaultComponentRef = React.useRef(null);
        var content = userProps.content, buttonKey = userProps.buttonKey, disabled = userProps.disabled, accessibilityLabel = userProps.accessibilityLabel, ariaLabel = userProps.ariaLabel, _d = userProps.componentRef, componentRef = _d === void 0 ? defaultComponentRef : _d, accessibilityPositionInSet = userProps.accessibilityPositionInSet, ariaPosInSet = userProps.ariaPosInSet, accessibilitySetSize = userProps.accessibilitySetSize, ariaSetSize = userProps.ariaSetSize, rest = __rest(userProps, ["content", "buttonKey", "disabled", "accessibilityLabel", "ariaLabel", "componentRef", "accessibilityPositionInSet", "ariaPosInSet", "accessibilitySetSize", "ariaSetSize"]);
        // Grabs the context information from RadioGroup (currently selected button and client's onChange callback)
        var info = React.useContext(RadioGroup_1.RadioGroupContext);
        var buttonRef = interactive_hooks_1.useViewCommandFocus(componentRef);
        /* We don't want to call the user's onChange multiple times on the same selection. */
        var changeSelection = function () {
            if (buttonKey != info.selectedKey) {
                info.onChange && info.onChange(buttonKey);
                info.updateSelectedButtonRef && componentRef && info.updateSelectedButtonRef(componentRef);
            }
        };
        /* We use the componentRef of the currently selected button to maintain the default tabbable
        element in a RadioGroup. Since the componentRef isn't generated until after initial render,
        we must update it once here. */
        React.useEffect(function () {
            if (buttonKey == info.selectedKey) {
                info.updateSelectedButtonRef && componentRef && info.updateSelectedButtonRef(componentRef);
            }
        }, []);
        // Ensure focus is placed on button after click
        var changeSelectionWithFocus = interactive_hooks_1.useOnPressWithFocus(componentRef, changeSelection);
        /* RadioButton changes selection when focus is moved between each RadioButton and on a click */
        var pressable = interactive_hooks_1.useAsPressable(__assign(__assign({}, rest), { onPress: changeSelectionWithFocus, onFocus: changeSelection }));
        // Used when creating accessibility properties in mergeSettings below
        var onAccessibilityAction = React.useCallback(function (event) {
            switch (event.nativeEvent.actionName) {
                case 'Select':
                    changeSelection();
                    break;
            }
        }, [info, buttonKey]);
        var state = __assign(__assign({}, pressable.state), { selected: info.selectedKey === userProps.buttonKey, disabled: disabled || false });
        // Grab the styling information from the userProps, referencing the state as well as the props.
        var styleProps = useStyling(userProps, function (override) { return state[override] || userProps[override]; });
        var slotProps = foundation_settings_1.mergeSettings(styleProps, {
            root: __assign(__assign(__assign(__assign({}, rest), { ref: buttonRef }), pressable.props), { accessibilityRole: 'radio', accessibilityLabel: (_a = accessibilityLabel !== null && accessibilityLabel !== void 0 ? accessibilityLabel : ariaLabel) !== null && _a !== void 0 ? _a : content, accessibilityState: { disabled: state.disabled, selected: state.selected }, accessibilityActions: [{ name: 'Select', label: RadioButton_settings_1.radioButtonSelectActionLabel }], accessibilityPositionInSet: (_b = accessibilityPositionInSet !== null && accessibilityPositionInSet !== void 0 ? accessibilityPositionInSet : ariaPosInSet) !== null && _b !== void 0 ? _b : info.buttonKeys.findIndex(function (x) { return x == buttonKey; }) + 1, accessibilitySetSize: (_c = accessibilitySetSize !== null && accessibilitySetSize !== void 0 ? accessibilitySetSize : ariaSetSize) !== null && _c !== void 0 ? _c : info.buttonKeys.length, onAccessibilityAction: onAccessibilityAction }),
            content: { children: content },
        });
        return { slotProps: slotProps };
    },
    render: function (Slots, _renderData) {
        var children = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            children[_i - 2] = arguments[_i];
        }
        return (foundation_composable_1.withSlots(Slots.root, null,
            foundation_composable_1.withSlots(Slots.button, null,
                foundation_composable_1.withSlots(Slots.innerCircle, null)),
            foundation_composable_1.withSlots(Slots.content, null),
            children));
    },
    settings: RadioButton_settings_1.settings,
    slots: {
        root: react_native_1.View,
        button: { slotType: react_native_1.View, filter: adapters_1.filterViewProps },
        innerCircle: { slotType: react_native_1.View, filter: adapters_1.filterViewProps },
        content: text_1.Text,
    },
    styles: {
        root: [],
        button: [tokens_1.borderTokens],
        innerCircle: [tokens_1.backgroundColorTokens],
        content: [tokens_1.foregroundColorTokens, tokens_1.textTokens, [{ source: 'textBorderColor', lookup: tokens_1.getPaletteFromTheme, target: 'borderColor' }]],
    },
});
exports.default = exports.RadioButton;
