// @fluentui-react-native/contextual-menu/lib-commonjs/SubmenuItem.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
/** @jsx withSlots */
var React = __importStar(require("react"));
var react_native_1 = require("react-native");
var SubmenuItem_types_1 = require("./SubmenuItem.types");
var foundation_compose_1 = require("@uifabricshared/foundation-compose");
var foundation_composable_1 = require("@uifabricshared/foundation-composable");
var text_1 = require("@fluentui-react-native/text");
var SubmenuItem_settings_1 = require("./SubmenuItem.settings");
var tokens_1 = require("@fluentui-react-native/tokens");
var foundation_settings_1 = require("@uifabricshared/foundation-settings");
var interactive_hooks_1 = require("@fluentui-react-native/interactive-hooks");
var react_native_svg_1 = require("react-native-svg");
var ContextualMenu_1 = require("./ContextualMenu");
var icon_1 = require("@fluentui-react-native/icon");
var interactive_hooks_2 = require("@fluentui-react-native/interactive-hooks");
exports.SubmenuItem = foundation_compose_1.compose({
    displayName: SubmenuItem_types_1.submenuItemName,
    usePrepareProps: function (userProps, useStyling) {
        var defaultComponentRef = React.useRef(null);
        var disabled = userProps.disabled, itemKey = userProps.itemKey, icon = userProps.icon, text = userProps.text, _a = userProps.accessibilityLabel, accessibilityLabel = _a === void 0 ? userProps.text : _a, onClick = userProps.onClick, testID = userProps.testID, _b = userProps.componentRef, componentRef = _b === void 0 ? defaultComponentRef : _b, rest = __rest(userProps, ["disabled", "itemKey", "icon", "text", "accessibilityLabel", "onClick", "testID", "componentRef"]);
        // Grabs the context information from Submenu (currently selected menuItem and client's onItemClick callback)
        var context = React.useContext(ContextualMenu_1.CMContext);
        var cmRef = interactive_hooks_1.useViewCommandFocus(componentRef);
        var onItemHoverIn = React.useCallback(function (e) {
            userProps.onHoverIn(e);
        }, []);
        var onItemClick = React.useCallback(function (e) {
            if (!disabled) {
                context === null || context === void 0 ? void 0 : context.onDismissMenu();
                (context === null || context === void 0 ? void 0 : context.dismissSubmenu) && context.dismissSubmenu();
                onClick ? onClick() : context === null || context === void 0 ? void 0 : context.onItemClick(itemKey);
                e.stopPropagation();
            }
        }, [context, disabled, itemKey, onClick]);
        var pressable = interactive_hooks_1.useAsPressable(__assign(__assign({}, rest), { onPress: onItemClick, onHoverIn: onItemHoverIn, delayHoverIn: 500 }));
        var _c = React.useState(false), submenuItemHovered = _c[0], setSubmenuItemHovered = _c[1];
        context.setSubmenuItemHovered = setSubmenuItemHovered;
        // set up state
        var state = __assign(__assign({}, pressable.state), { selected: context.selectedKey === userProps.itemKey, disabled: userProps.disabled, content: !!text, icon: !!icon, submenuItemHovered: submenuItemHovered });
        var onMouseEnter = React.useCallback(function (e) {
            setSubmenuItemHovered(true);
            pressable.props.onMouseEnter && pressable.props.onMouseEnter(e);
            e.stopPropagation();
        }, [pressable, setSubmenuItemHovered]);
        /*
         * SubmenuItem launches the submenu onMouseEnter event. For keyboarding, submenu should be launched with Spacebar, Enter, or right arrow.
         */
        var onKeyUp = interactive_hooks_1.useKeyCallback(onMouseEnter, ' ', 'Enter', 'ArrowRight');
        // grab the styling information, referencing the state as well as the props
        var styleProps = useStyling(userProps, function (override) { return state[override] || userProps[override]; });
        // create the merged slot props
        var slotProps = foundation_settings_1.mergeSettings(styleProps, {
            root: __assign(__assign({}, pressable.props), { ref: cmRef, onKeyUp: onKeyUp,
                onMouseEnter: onMouseEnter,
                accessibilityLabel: accessibilityLabel }),
            content: { children: text, testID: testID },
            icon: interactive_hooks_2.createIconProps(icon),
            chevron: {},
        });
        return { slotProps: slotProps, state: state };
    },
    settings: SubmenuItem_settings_1.settings,
    render: function (Slots, renderData) {
        var children = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            children[_i - 2] = arguments[_i];
        }
        var xml = "\n    <svg width=\"12\" height=\"12\" viewBox=\"0 0 2048 2048\" >\n      <path class='OfficeIconColors_HighContrast' fill='currentColor' d='M 743 1767 l -121 -121 l 708 -707 l -708 -708 l 121 -121 l 828 829 z' />\n      <path class='OfficeIconColors_m22' fill='currentColor' d='M 743 1767 l -121 -121 l 708 -707 l -708 -708 l 121 -121 l 828 829 z' />\n    </svg>";
        // We shouldn't have to specify the source prop on Slots.icon, here, but we need another drop from @uifabricshared
        return (foundation_composable_1.withSlots(Slots.root, null,
            foundation_composable_1.withSlots(Slots.leftstack, null,
                renderData.state.icon && foundation_composable_1.withSlots(Slots.icon, null),
                renderData.state.content && foundation_composable_1.withSlots(Slots.content, null),
                children),
            foundation_composable_1.withSlots(Slots.rightstack, null,
                foundation_composable_1.withSlots(Slots.chevron, { xml: xml }))));
    },
    slots: {
        root: react_native_1.View,
        leftstack: { slotType: react_native_1.View },
        icon: { slotType: icon_1.Icon },
        content: text_1.Text,
        rightstack: { slotType: react_native_1.View },
        chevron: react_native_svg_1.SvgXml,
    },
    styles: {
        root: [tokens_1.backgroundColorTokens, tokens_1.borderTokens],
        leftstack: [],
        icon: [{ source: 'iconColor', lookup: tokens_1.getPaletteFromTheme, target: 'color' }],
        content: [tokens_1.textTokens, tokens_1.foregroundColorTokens],
        rightstack: [],
        chevron: [{ source: 'chevronColor', lookup: tokens_1.getPaletteFromTheme, target: 'color' }],
    },
});
exports.default = exports.SubmenuItem;
